mrmustard/physics/gaussian_integrals.py

```diff
@@ -259,216 +160,307 @@ def reorder_abc(Abc: tuple, order: Sequence[int]):
+    if len(idx_z) != len(idx_zconj):
         raise ValueError(
-            f"idx1 and idx2 must have the same length, got {len(idx1)} and {len(idx2)}"
+            f"idx1 and idx2 must have the same length, got {len(idx_z)} and {len(idx_zconj)}"
+        )
+    A, b, c = Abc
+    c = math.astensor(c)
+    # assuming c is batched accordingly
+    batched = len(A.shape) == 3 and len(b.shape) == 2 and len(c.shape) > 0
+    if not batched:
+        A = math.atleast_3d(A)
+        b = math.atleast_2d(b)
+        c = math.atleast_1d(c)
+    if not A.shape[0] == b.shape[0] == c.shape[0]:
+        raise ValueError(
+            f"Batch size mismatch: got {A.shape[0]} for A, {b.shape[0]} for b and {c.shape[0]} for c."
         )
-    n = len(idx1)
-    idx = tuple(idx1) + tuple(idx2)
-    if not idx:
-        raise NotImplementedError
 
-    A = math.block_diag(math.cast(A1, "complex128"), math.cast(A2, "complex128"))
-    b = math.concat([b1, b2], axis=-1)
-    c = math.reshape(math.outer(c1, c2), c1.shape + c2.shape)
+    (_, n_plus_N, _) = A.shape
+    if b.shape[-1] != A.shape[-1]:
+        raise ValueError(f"A and b must have compatible shapes, got {A.shape} and {b.shape}")
+    N = len(c.shape[1:])  # number of beta variables
+    n = n_plus_N - N  # number of z variables
+    m = len(idx_z)  # number of pairs to integrate over
+    idx = tuple(idx_z) + tuple(idx_zconj)
+    if any(i >= n for i in idx):
+        raise ValueError(
+            f"Indices must be less than {n}, got {tuple(i for i in idx_z if i >= n)} and {tuple(i for i in idx_zconj if i >= n)}"
+        )
 
-    not_idx = tuple(i for i in range(A.shape[-1]) if i not in idx)
+    if len(idx) == 0:
+        if not batched:
+            return A[0], b[0], c[0]
+        return A, b, c
 
-    I = math.eye(n, dtype=A.dtype)
-    Z = math.zeros((n, n), dtype=A.dtype)
-    X = math.block([[Z, I], [I, Z]])
+    not_idx = tuple(i for i in range(n_plus_N) if i not in idx)
+    eye = math.eye(m, dtype=A.dtype)
+
+    eye = math.eye(m, dtype=A.dtype)
+    Z = math.zeros((m, m), dtype=A.dtype)
+    X = math.block([[Z, eye], [eye, Z]])
     M = math.gather(math.gather(A, idx, axis=-1), idx, axis=-2) + X * measure
     bM = math.gather(b, idx, axis=-1)
 
-    c_post = (
-        c * math.sqrt((-1) ** n / math.det(M)) * math.exp(-0.5 * math.sum(bM * math.solve(M, bM)))
-    )
-
-    if math.asnumpy(not_idx).shape != (0,):
-        D = math.gather(math.gather(A, idx, axis=-1), not_idx, axis=-2)
-        R = math.gather(math.gather(A, not_idx, axis=-1), not_idx, axis=-2)
-        bR = math.gather(b, not_idx, axis=-1)
-        A_post = R - math.matmul(D, math.inv(M), math.transpose(D))
-        b_post = bR - math.matvec(D, math.solve(M, bM))
+    determinant = math.det(M)
+    D = math.gather(math.gather(A, idx, axis=-1), not_idx, axis=-2)
+    R = math.gather(math.gather(A, not_idx, axis=-1), not_idx, axis=-2)
+    bR = math.gather(b, not_idx, axis=-1)
+    if np.all(math.abs(determinant)) > 1e-12:
```

Comment:
Should this be a `math.allclose`?

Reply:
no, I want to test if they are all not zero